### PR TITLE
RavenDB-25281 - Corax 2.0 stabilization: server, PAL, Voron & Sparrow review fixes

### PR DESCRIPTION
**Pull request by ayende**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-25281

It contains fixes for the following tests:

https://issues.hibernatingrhinos.com/issue/RavenDB-26806/FastTests.Sharding.Queries.BasicShardedMapReduceStreamingTests.AutoMapReduceWithOrderByoptions-DatabaseMode-Sharded
https://issues.hibernatingrhinos.com/issue/RavenDB-26807/FastTests.Sharding.Queries.BasicShardedMapReduceStreamingTests.AutoMapReduceWithOrderByoptions-DatabaseMode-Sharded

### Additional description

Server-side stabilization fixes extracted from the Corax 2.0 review, kept independent of the Corax engine changes so they can be reviewed and merged on their own. All fixes sit directly on top of `v8.0`.

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [x] Yes. Please list the affected platforms.

The PAL fixes (`madvise`, double-`fd`-close, `eventfd_read` `EINTR`, native binaries) are POSIX-specific (Linux / macOS).

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.

The pager prefetch / `madvise` changes touch the native I/O path.

### Is there any existing behavior change of other features due to this change?

- [x] No

### UI work

- [x] No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/ETL/EtlProcessStatistics.cs [hunk 1]
```diff
@@ -148,7 +148,7 @@ public void RecordItemTransformationError(Exception e, string documentId)
             BatchErrors++;
         }
 
-        public void RecordItemLoadError(string error, string documentId, int count = 1)
+        public void RecordItemLoadError(string error, string documentId, int count = 1, TaskErrorStep step = TaskErrorStep.Load)
```
Comment (Lwiel) [inline src/Raven.Server/Documents/ETL/EtlProcessStatistics.cs:149]: Maybe let's add a debug assert that only actual load steps are passed here, because this allows to pass any step, even non-load ones

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs [hunk 2]
```diff
@@ -212,19 +212,21 @@ private async Task WriteServerMetricsAsync(MetricsProvider provider, Stream resp
...
-
-                    // AI Tasks
-                    WriteCounterWithHelp(writer, "Number of AI tasks", "server_ai_tasks_count", serverMetrics.AiTasks.Count);
-                    WriteCounterWithHelp(writer, "Number of AI task errors", "server_ai_tasks_errors_count", serverMetrics.AiTasks.ErrorsCount);
-                    WriteCounterWithHelp(writer, "Number of healthy AI tasks", "server_ai_tasks_healthy_count", serverMetrics.AiTasks.HealthyTasksCount);
-                    WriteCounterWithHelp(writer, "Number of impaired AI tasks", "server_ai_tasks_impaired_count", serverMetrics.AiTasks.ImpairedTasksCount);
-                    WriteCounterWithHelp(writer, "Number of failed AI tasks", "server_ai_tasks_failed_count", serverMetrics.AiTasks.FailedTasksCount);
+                    // ETLs — current-state counts that fall when a database unloads or an ETL/error is removed,
+                    // so they must be gauges: a Prometheus counter that decreases is read as a reset and corrupts
+                    // rate()/increase(). Matches the per-database ETL metrics, which already use gauges.
```
Comment (Lwiel) [inline src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: I don't think we need these comments, but the change itself is good

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs [hunk 3]
```diff
@@ -212,19 +212,21 @@ private async Task WriteServerMetricsAsync(MetricsProvider provider, Stream resp
...
+                    WriteGaugeWithHelp(writer, "Number of ETLs", "server_etls_count", serverMetrics.Etls.Count);
+                    WriteGaugeWithHelp(writer, "Number of ETL errors", "server_etls_errors_count", serverMetrics.Etls.ErrorsCount);
+                    WriteGaugeWithHelp(writer, "Number of healthy ETLs", "server_etls_healthy_count", serverMetrics.Etls.HealthyEtlsCount);
+                    WriteGaugeWithHelp(writer, "Number of impaired ETLs", "server_etls_impaired_count", serverMetrics.Etls.ImpairedEtlsCount);
+                    WriteGaugeWithHelp(writer, "Number of failed ETLs", "server_etls_failed_count", serverMetrics.Etls.FailedEtlsCount);
+
+                    // AI Tasks — same rationale as the ETL counts above (current-state, can decrease → gauges).
```
Comment (Lwiel) [inline src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: Same here

File: src/Sparrow.Server/Binary/PtrBitVector.cs [hunk 4]
```diff
@@ -108,21 +109,34 @@ public bool AllEmpty()
...
             }
 
             return -1;
+
+            static int ResolveIndex(int byteOffset, ulong value, int count)
+            {
+                int idx = byteOffset * 8 + BitOperations.LeadingZeroCount(
+                    // this is needed because we set on byte boundaries, but read using ulong
+                    BinaryPrimitives.ReverseEndianness(value));
+                return idx < count ? idx : -1; // avoid returning an index that is greater than the count (e.g. if the last byte is partially used)
```
[the remark this answers is not in the document]
Comment (ayende) [reply src/Sparrow.Server/Binary/PtrBitVector.cs:138]: That is not a concern, we are not supporting big endian

File: src/Sparrow.Server/Binary/PtrBitVector.cs [hunk 5]
```diff
@@ -108,21 +109,34 @@ public bool AllEmpty()
...
+            int lengthInBytes = (Count + 7) / 8; // Iterate in byte units
+
             int i = 0;
-            for (; i < Count / 8; i+=8)
+            for (; i + sizeof(ulong) <= lengthInBytes; i += sizeof(ulong))
             {
-                var l = *(ulong*)(Bits + i);
-                if (l!= 0)
-                    return BitOperations.LeadingZeroCount(l) + i * 8;
+                ulong l = *(ulong*)(Bits + i);
```
[the remark this answers is not in the document]
Comment (ayende) [reply src/Sparrow.Server/Binary/PtrBitVector.cs]: fixed

File: src/Raven.Pal/src/posix/ioring.c [hunk 6]
```diff
@@ -390,7 +395,12 @@ wait_for_work_completion(struct handle *handle_ptr, struct submittion *submittio
         return FAIL_IO_RING_WRITE;
     }
     eventfd_t v;
-    eventfd_read(submittion->notifyfd, &v);
+    int read_rc;
+    do
+    {
+        read_rc = eventfd_read(submittion->notifyfd, &v);
+    } while (read_rc == -1 && errno == EINTR);
```
[the remark this answers is not in the document]
Comment (ayende) [reply src/Raven.Pal/src/posix/ioring.c:402]: fixed

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: Moved that change from this PR entirely

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: Moved that change from this PR entirely

File: src/Raven.Server/Documents/ETL/EtlProcessStatistics.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/ETL/EtlProcessStatistics.cs:149]: agreed

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs [hunk 7]
```diff
@@ -212,19 +212,21 @@ private async Task WriteServerMetricsAsync(MetricsProvider provider, Stream resp
...
-                    WriteCounterWithHelp(writer, "Number of AI task errors", "server_ai_tasks_errors_count", serverMetrics.AiTasks.ErrorsCount);
-                    WriteCounterWithHelp(writer, "Number of healthy AI tasks", "server_ai_tasks_healthy_count", serverMetrics.AiTasks.HealthyTasksCount);
-                    WriteCounterWithHelp(writer, "Number of impaired AI tasks", "server_ai_tasks_impaired_count", serverMetrics.AiTasks.ImpairedTasksCount);
-                    WriteCounterWithHelp(writer, "Number of failed AI tasks", "server_ai_tasks_failed_count", serverMetrics.AiTasks.FailedTasksCount);
+                    // ETLs — current-state counts that fall when a database unloads or an ETL/error is removed,
+                    // so they must be gauges: a Prometheus counter that decreases is read as a reset and corrupts
+                    // rate()/increase(). Matches the per-database ETL metrics, which already use gauges.
+                    WriteGaugeWithHelp(writer, "Number of ETLs", "server_etls_count", serverMetrics.Etls.Count);
+                    WriteGaugeWithHelp(writer, "Number of ETL errors", "server_etls_errors_count", serverMetrics.Etls.ErrorsCount);
```
[the remark this answers is not in the document]
Comment (ayende) [reply src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: removed that to https://issues.ravendb.net/issue/RavenDB-27019

**Review by maciejaszyk — APPROVED** [review]

**Review by redknightlois — COMMENTED** [review]
## Overview

Multi-system stabilization fixes extracted from the Corax 2.0 review, applied directly to `v8.0`: a faulty `madvise(MADV_WILLNEED)` return-value check that silently disabled Linux prefetch, POSIX `eventfd_read` hardened against `EINTR` signals, a double-`fd`-close guard in the PAL pager, a `PtrBitVector.FirstSetBit` rewrite with correct endianness normalization and unaligned reads, stale value-copy reads fixed in `LockFreeRingBuffer` atomics, the current ETL load step propagated into per-item error records, leaked `TaskCompletionSource` instances cancelled on aborted Rachis transactions, cluster-state churn reduced during map-reduce output-collection prefix deletes, the sharded footer-stats endpoint hardened against unreachable replicas, and a safe fallback for `IdentityPartsSeparator` during early database startup. PAL native binaries bumped from 7.0.888 to 7.0.889.

## Changes

- **native-pal:** Fixed inverted `madvise(MADV_WILLNEED)` check in `rvn_prefetch_ranges` (returned FAIL on success and silently swallowed errors); `eventfd_read` calls in `do_ring_work` and `wait_for_work_completion` now retry on `EINTR` and check the return code; `rvn_init_pager` and `rvn_increase_pager_size` guard against double-`close(fd)` when `_open_pager_file` already owns the descriptor.
- **voron-prefetch:** `GetSegmentState` bounds check corrected — the old comparison mixed segment numbers with byte-array length, silently under-utilizing the prefetch table for large files. `SetSegmentState` now mutates the array element via `ref` instead of copy-then-store.
- **sparrow-primitives:** `PtrBitVector.FirstSetBit` rewritten with `Unsafe.ReadUnaligned<ulong>`, `BinaryPrimitives.ReverseEndianness` normalization, and an `idx < count` clamp to exclude padding bits in every boundary case. `LockFreeRingBuffer` Count/IsEmpty/IsFull read cell sequences via `ref` into the array rather than stale value-copies.
- **etl / ai-etl:** `EtlProcessStatistics.RecordItemLoadError` accepts an optional `step` parameter; `EmbeddingsGenerationTask` and `GenAiTask` use try/catch to capture the failing phase and attribute per-item errors correctly; `EtlProcess.Run` snapshots `FallbackTime` via pattern-match to close a TOCTOU window against concurrent nullification.
- **rachis-tx-notify:** Four `OnDispose` handlers on `LowLevelTransaction` now call `RachisLogIndexNotifications.CancelTask` when the transaction is aborted, preventing leaked `TaskCompletionSource` instances. `AddTask` uses `TryAdd` (first-writer-wins) to handle duplicate index registration within a single `Apply`.
- **server-resilience:** Map-reduce prefix deletes now batch the cluster-state update (one call after the loop instead of one per iteration). Sharded footer stats tolerate unreachable replicas via try/catch with cancellation-respecting filter. SNMP database monitoring catches `DatabaseDisabledException` in the fire-and-forget continuation. `DocumentDatabase.IdentityPartsSeparator` falls back to `Constants.Identities.DefaultSeparator` instead of throwing when `ClientState` is unavailable during early startup.

<details><summary>Show a summary per file</summary>

| File | Description |
|---|---|
| `src/Raven.Pal/src/posix/virtualmemory.c` | Fixed inverted `madvise` return-value check |
| `src/Raven.Pal/src/posix/ioring.c` | EINTR-retry loops and return-code checks on `eventfd_read` |
| `src/Raven.Pal/src/posix/pager.c` | `fd = -1` guard against double-close in init and resize paths |
| `src/Voron/Impl/Paging/Pager.cs` | Fixed segment-index bounds; `SetSegmentState` ref-mutation |
| `src/Sparrow.Server/Binary/PtrBitVector.cs` | Endianness-aware unaligned `FirstSetBit` with padding clamp |
| `src/Sparrow/Collections/LockFreeRingBuffer.cs` | `ref`-based `Volatile.Read` on shared cell sequences |
| `src/Raven.Server/Documents/ETL/EtlProcess.cs` | TOCTOU-safe `FallbackTime` snapshot, null fallback in `WaitAny` |
| `src/Raven.Server/Documents/ETL/EtlProcessStatistics.cs` | Optional `step` parameter on `RecordItemLoadError` |
| `src/Raven.Server/Documents/ETL/Providers/AI/Embeddings/EmbeddingsGenerationTask.cs` | Try/catch captures failed load step for error attribution |
| `src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs` | Try/catch + dynamic `LoadErrorStep` for per-item step tagging |
| `src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiBatchPatchCommand.cs` | Explicit `TaskErrorStep.Persistence` on batch errors |
| `src/Raven.Server/ServerWide/RachisLogIndexNotifications.cs` | `TryAdd` dedup + new `CancelTask` method |
| `src/Raven.Server/ServerWide/ClusterStateMachine.cs` | `CancelTask` called on aborted transactions in 4 handlers |
| `src/Raven.Server/Documents/DocumentDatabase.cs` | Default separator fallback when `ClientState` unavailable |
| `src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionActions.cs` | Batched cluster-state update after prefix-delete loop |
| `src/Raven.Server/Documents/Sharding/Handlers/Processors/Studio/ShardedStudioStatsHandlerProcessorForGetFooterStats.cs` | Per-replica fault tolerance in footer stats |
| `src/Raven.Server/Monitoring/Snmp/SnmpDatabase.cs` | `DatabaseDisabledException` catch in SNMP monitoring |
| `test/FastTests/Sparrow/PtrBitVector.cs` | New tests for `FirstSetBit` boundary cases |

</details>

## Cross-cutting observations

- Commit `c418c003d` body includes the claim "Prometheus: emit server ETL/AI task counts as gauges (they decrease on unload/removal)" — no Prometheus-related changes appear in the diff, suggesting the claim refers to work committed separately or the body was carried over from a larger branch.

## Tests

- New `PtrBitVectorTests` (FastTests/Sparrow) covers `FirstSetBit` consistency with the indexer, lowest-set-index correctness (bits 5, 40, 63), padding-bit exclusion (Count=5 with bit 5 set → -1), and the fast-path padding clamp (Count=60 with bit 61 set → -1). Uses `[RavenTheory]`/`[RavenFact]` with `RavenTestCategory.Memory`, inheriting from `NoDisposalNeeded`.

## Backward compatibility

- Non-breaking change. All PAL fixes are POSIX-native and transparent to the managed API surface. `RecordItemLoadError`'s new `step` parameter defaults to `TaskErrorStep.Load`, preserving existing call-site behaviour. PAL version bump (70888→70889) is wire-compatible at the P/Invoke boundary. `IdentityPartsSeparator` fallback only activates during the early-startup window when `ClientState` is unavailable and no documents exist.

## Coverage

### primitives.md

| Dimension | Status |
|---|---|
| Native error propagation | Covered — ioring EINTR retry, madvise fix, eventfd_read error check reviewed |
| Concurrency | Covered — LockFreeRingBuffer ref-atomicity fix reviewed |
| Lifetime & ownership | Covered — fd ownership contract in pager.c reviewed |
| On-disk / ABI compatibility | Covered — PAL_VER (70889) consistent across pal.ver, rvngetpalver.c, Pal.cs, Directory.Packages.props |
| Performance discipline | N/A — no `PERF`-marked code in the changed paths |

### voron.md

| Dimension | Status |
|---|---|
| Concurrency | Covered — pre-existing tolerant-read model unchanged; ref-mutation equivalent to copy-then-store |
| Lifetime & ownership | N/A — no native or Voron resource allocations in the diff |
| On-disk compatibility | N/A — no page-format or schema-version changes |
| Performance discipline | Covered — ref-mutation eliminates one bounds check; inline-method body shrank, not grew |

- `server.md`: Server pack has no required dimensions — rigor from Phase 1a baseline and always-report lines; no triggers hit.
- `ai-embeddings.md`: AI embeddings pack has no required dimensions — rigor from Phase 1a baseline and always-report lines; no triggers hit (no JS-exposed helpers, argument parsing, or script-option dispatch touched).

## Verdict

No blocking issues — looks good to merge.

---

<details><summary>Considered &amp; discarded (3) — audit trail</summary>

Medium/high-confidence candidates the scan stage raised that a later stage dropped. Each row is the reason it was discarded — re-open it if the rationale looks thin.

| Finding | Where | Severity · confidence | Dropped at | Why |
|---|---|---|---|---|
| bare Claude session URL in commit body | :0 | Concern · High | judge | The commit body is self-contained — bug, cause, impact, and fix are all inline; a domain-bearing URL with descriptive prefix matches none of the rule's four bare-shapes (ordinals, short codes, file:line anchor, bare pronouns).; The consequence is negligible — a dead-link footnote in an otherwise self-contained commit message. No reader's understanding of the change depends on the URL; the commit body fully explains the bug, cause, impact, and fix inline. Absence of material harm.; Rule (Phase 1a bare short codes) does not genuinely apply: the cited shapes are opaque short codes (V1,R3,F2) and bare ordinals — a labeled, domain-bearing URL with a descriptive prefix (Claude-Session:) is not bare, and the commit body is fully self-contained without it. |
| fd ownership contract asserted in comment but callee was not changed | src/Raven.Pal/src/posix/pager.c:302 | Concern · Medium | judge | _open_pager_file already closes fd on every failure path (all goto Error reach close(fd)) and stores it on every success path; the comment accurately describes existing behavior, not a new unfulfilled contract.; The finding's claimed defect (fd leak from callee not matching the comment's contract) is refuted: both advocates show _open_pager_file already closes fd on every failure path. The comment documents existing behaviour, not a missing change. No consequence.; Rule (primitives.md lifetime/ownership) does not genuinely apply: _open_pager_file already closes fd on every error path (goto Error → close(fd)) and stores it on every success path — the comment documents existing behavior, not an unfulfilled contract. No leak exists. |
| LoadErrorStep not reset per LoadInternal — cross-batch state leakage | src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs:179 | Concern · Low | judge | EnterLoadStep at EtlProcess.cs:536 unconditionally sets LoadErrorStep=ModelInference before HandleItemError runs; HandleItemError at GenAiTask.cs:374 executes inside that using scope, so LoadErrorStep is always ModelInference when read — the scanner's stale-Persistence walkthrough is factually impossible.; The finding's claimed mechanism (cross-batch LoadErrorStep leakage) is refuted: EnterLoadStep unconditionally overwrites the property to ModelInference before HandleItemError runs, on every LoadInternal invocation. The stale value from batch 1's catch is never observed by any reader.; Rule (mutable instance state not reset at method entry) does not genuinely apply: EnterLoadStep unconditionally overwrites LoadErrorStep to ModelInference before any reader runs within the using block — the property IS reset at the point it matters. The scanner's walkthrough depends on EnterLoadStep not working, but it demonstrably does. |

</details>

File: src/Sparrow/Collections/LockFreeRingBuffer.cs [hunk 8]
```diff
@@ -127,18 +127,16 @@ public bool TryDequeue(out T item)
...
             get
             {
+                // managed refs into the array so Volatile.Read reads directly from there
                 long currentEnqueuePos = Volatile.Read(ref _positions.Enqueue);
-                Cell enqueueCell = _buffer[currentEnqueuePos & _bufferMask];
+                ref Cell enqueueCell = ref _buffer[currentEnqueuePos & _bufferMask];
 
                 long currentDequeuePos = Volatile.Read(ref _positions.Dequeue);
-                Cell dequeueCell = _buffer[currentDequeuePos & _bufferMask];
+                ref Cell dequeueCell = ref _buffer[currentDequeuePos & _bufferMask];
```
Comment (redknightlois) [inline src/Sparrow/Collections/LockFreeRingBuffer.cs:139]: This is a quite neat improvement

**Review by redknightlois — APPROVED** [review]

**Review by arekpalinski — DISMISSED** [review]

File: test/FastTests/Sharding/Queries/RavenDB_MapReduceStreamMergeArenaReset.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
+using System.Collections.Generic;
+using System.Linq;
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Documents.Linq;
+using Tests.Infrastructure;
+using Xunit;
+
+namespace FastTests.Sharding.Queries
+{
+    public class RavenDB_MapReduceStreamMergeArenaReset : RavenTestBase
```
Comment (ayende) [inline test/FastTests/Sharding/Queries/RavenDB_MapReduceStreamMergeArenaReset.cs:10]: @karmeli87 - Look at this fix please. I added that to v7.2 and v8.0 - do you think we need it for v6.2 as well?

**Comment by arekpalinski** [comment]
I did the rebase and solved conflicts here.

**Review by arekpalinski — APPROVED** [review]
